Optimize Active Record batching for whole table iterations

fatkodima opened the pull request:
Reference #44945.

Tested on a table with 10M records.

```sql
CREATE TABLE users (id bigserial PRIMARY KEY, val integer DEFAULT 0);
INSERT INTO users SELECT i FROM generate_series(1, 10000000) AS i;
```

### Selecting

```ruby
start = Process.clock_gettime(Process::CLOCK_MONOTONIC)

count = 0
User.in_batches do |batch|
  count += batch.count
end

puts "Count = #{count}"
elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
puts "Elapsed: #{elapsed}s"
```

Before - `253s`, after: `30s` 🔥.

**Single batch plan before**:
```
explain analyze SELECT COUNT(*) FROM "users" WHERE "users"."id" IN (1, 2, 3, 4, 5....

 Aggregate  (cost=4387.00..4387.01 rows=1 width=8) (actual time=1.767..1.768 rows=1 loops=1)
   ->  Index Only Scan using users_pkey on users  (cost=0.43..4384.50 rows=1000 width=0) (actual time=0.049..1.668 rows=1000 loops=1)
         Index Cond: (id = ANY ('{1,2,3,4,5....
         Heap Fetches: 0
 Planning Time: 1.616 ms
 Execution Time: 1.803 ms
```

**Single batch plan after**:
```
explain analyze SELECT COUNT(*) FROM (SELECT 1 AS one FROM "users" WHERE "users"."id" <= 1000 ORDER BY "users"."id" ASC LIMIT 1000) subquery_for_count;
                                                                QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=44.34..44.35 rows=1 width=8) (actual time=0.404..0.405 rows=1 loops=1)
   ->  Limit  (cost=0.43..31.84 rows=1000 width=12) (actual time=0.036..0.318 rows=1000 loops=1)
         ->  Index Only Scan using users_pkey on users  (cost=0.43..36.58 rows=1151 width=12) (actual time=0.031..0.184 rows=1000 loops=1)
               Index Cond: (id <= 1000)
               Heap Fetches: 0
 Planning Time: 0.137 ms
 Execution Time: 0.436 ms
(7 rows)
```

### Updating

```ruby
start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
User.in_batches.update_all("val = val + 1")
elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
puts "Elapsed: #{elapsed}s"
```

Before - `288s`, after: `124s` 🔥.

**Single batch plan before**:
```
explain analyze UPDATE "users" SET val = val + 1 WHERE "users"."id" IN (1001, 1002, 1003, ..., 2000)

 Update on users  (cost=0.43..4396.00 rows=0 width=0) (actual time=27.615..27.616 rows=0 loops=1)
   ->  Index Scan using users_pkey on users  (cost=0.43..4396.00 rows=1000 width=10) (actual time=0.054..4.064 rows=1000 loops=1)
         Index Cond: (id = ANY ('{1001,1002,1003,1004,...
 Planning Time: 1.154 ms
 Execution Time: 9.645 ms
```

**Single batch plan after**:
```
explain analyze UPDATE "users" SET val = val + 1 WHERE "users"."id" IN (SELECT "users"."id" FROM "users" WHERE "users"."id" > 1000 AND "users"."id" <= 2000 ORDER BY "users"."id" ASC LIMIT 1000)

 Update on users  (cost=50.67..8411.24 rows=0 width=0) (actual time=13.461..13.463 rows=0 loops=1)
   ->  Nested Loop  (cost=50.67..8411.24 rows=1000 width=42) (actual time=2.040..5.586 rows=1000 loops=1)
         ->  HashAggregate  (cost=50.24..60.24 rows=1000 width=40) (actual time=2.012..2.345 rows=1000 loops=1)
               Group Key: "ANY_subquery".id
               Batches: 1  Memory Usage: 193kB
               ->  Subquery Scan on "ANY_subquery"  (cost=0.43..47.74 rows=1000 width=40) (actual time=0.023..1.675 rows=1000 loops=1)
                     ->  Limit  (cost=0.43..37.74 rows=1000 width=8) (actual time=0.019..1.417 rows=1000 loops=1)
                           ->  Index Only Scan using users_pkey on users users_1  (cost=0.43..43.56 rows=1156 width=8) (actual time=0.018..1.284 rows=1000 loops=1)
                                 Index Cond: ((id > 1000) AND (id <= 2000))
                                 Heap Fetches: 2097
         ->  Index Scan using users_pkey on users  (cost=0.43..8.35 rows=1 width=18) (actual time=0.003..0.003 rows=1 loops=1000)
               Index Cond: (id = "ANY_subquery".id)
 Planning Time: 0.216 ms
 Execution Time: 13.555 ms
```

### Deleting

```ruby
start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
User.in_batches.delete_all
elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
puts "Elapsed: #{elapsed}s"
```

Before - `268s`, after: `83s` 🔥.

cc @byroot @nvasilevski @simi @etiennebarrie

fatkodima committed:
Optimize Active Record batching for whole table iterations

byroot commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -233,6 +233,14 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
           ids = records.map(&:id)
           yielded_relation = where(primary_key => ids)
           yielded_relation.load_records(records)
+        elsif self == klass.unscoped.all
```
We can probably move that scope out of the loop.

byroot replied:
Sorry `scope` -> `check`

fatkodima replied:
If we move it out of the loop, we need to have a condition like `range_batching_possible = !load && self == klass.unscoped.all` and use this variable then. We need `!load`, because otherwise the comparison will load records for association relations https://github.com/rails/rails/blob/7669a4a9757bff49bfc37d772886f59a90dd0434/activerecord/lib/active_record/association_relation.rb#L14-L16

Leaving it as is seems just cleaner, imo?

byroot replied:
Hum, I'm just a bit annoyed to have to instantiate two relations on every loop. Maybe we just cache `klass.unscoped.all`?

That `==` implementation is scary too. I fear we might trigger some awful query by accident, but I guess we have some tests that would fail, right?

byroot reviewed (commented):
A test would be nice.

Other than that the condition is simple enough, that I'm not concerned about increased maintnance.

fatkodima commented:
Added test cases.

matthewd commented:
Given it's (so measurably) fast enough to matter, it does feel to me like we should be exposing this as a [rarely used, but available] option to override this heuristic... if the only condition in play is a `archived_at: nil` (and only a tiny fraction of your records are archived), it may well still make sense to opt in to this approach?

fatkodima commented:
> it may well still make sense to opt in to this approach

I think yes, we just need a good (highlevel?) name for this option 😅

simi commented:
I'll test this on real data and report back. Btw. @fatkodima it is good practice on PostgreSQL after each big insert of data to run ANALYZE on table to get proper statistics for query planner.

simi commented:
> > it is good practice on PostgreSQL after each big insert of data to run ANALYZE on table to get proper statistics for query planner
> 
> Sure, but we index only scan the primary key index here, so this should not make a difference.

```sql
DROP TABLE users1;
CREATE TABLE users1 (id bigserial PRIMARY KEY, val integer DEFAULT 0);
INSERT INTO users1 SELECT i FROM generate_series(1, 10000000) AS i;
EXPLAIN SELECT id FROM users1 WHERE id IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
ANALYZE users1;
EXPLAIN SELECT id FROM users1 WHERE id IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
```

:thinking: I get different costs for both queries. I'm not sure how that is affecting your numbers, but I was just about to mention it is good idea to start with analyzed data.

fatkodima commented:
@simi Were you able to test this on real data - https://github.com/rails/rails/pull/45414#issuecomment-1161590945 ?

simi commented:
> > I can check again later on EXPLAIN ANALYZE at least in transaction and rollback on my testing data to get better insight.
> 
> +1 Even though the query plan may look scarier, the real execution time can be better.
> 
> It would be great, if you can suggest some sort of a reproduction script (to be able to test locally), that proves the new approach can be slower.

I'm trying to point to the problem that `delete_all` on batch is being transformed in different way to `DELETE` query (using subquery) with this change. Good next step would be to check `update_all`.

> This list is often very large (1000 by default), which can lead to inefficient queries and truncated sql queries in logs and monitoring tools. I'm sure everyone saw that monstrous queries of the half of the screen in size smile

This was the original description and motivation for this change. Maybe there is actual reason for the current implementation. I can try to get more testing results if welcomed. :pray:

fatkodima commented:
> I'm trying to point to the problem that delete_all on batch is being transformed in different way to DELETE query (using subquery) with this change. Good next step would be to check update_all.

I measured `delete_all/update_all` (see PR description), but yes, I used only the default batch size of 1000. I will try to test with bigger tables and different batch sizes.  

> I can try to get more testing results if welcomed. 🙏

Sure ❤️

simi commented:
> > I'm trying to point to the problem that delete_all on batch is being transformed in different way to DELETE query (using subquery) with this change. Good next step would be to check update_all.
> 
> I measured `delete_all/update_all` (see PR description), but yes, I used only the default batch size of 1000. I will try to test with bigger tables and different batch sizes.

I'm related to PostgreSQL only. I'm able to evaluate only query plans. Would you mind to share query plans as well for your testing in PR description?

fatkodima commented:
Updated description with query plans.

Ok, tested with the same 10m table and 100k batch size this time.

Select: `238s` vs `7s`. 🔥 🔥 🔥 
Update: `204s` vs `104s`. 🔥 
Delete: `172s` vs `323s` 😱.

So the new approach can be slower.

But 100k is kinda unrealistic batch size, from my experience.

Tested deletes with other batch sizes:
* 10k: `114s` vs `69s`
* 30k: `129s` vs `69s`
* 50k: `136s` vs `61s`

fatkodima commented:
@byroot 
1. Wdyt about Matthew's suggestion at https://github.com/rails/rails/pull/45414#issuecomment-1161489698?

2. And given this PR can make (while very unlikely) the batching slower (see the last comment), should this PR's optimization be applied by default or must be enabled by the option from point 1?

byroot commented:
> 1. Wdyt about Matthew's suggestion at

Yeah, I think having a param to manually opt-in to this bahavior on arbitrary scope would make perfect sense.

> should this PR's optimization be applied by default 

I think it should be used by default for empty scopes yes.

fatkodima commented:
@byroot @matthewd Updated. Please, take another look.

byroot commented:
> Have you checked my finding

Yes, and unless I missed something, it's only less performance on really huge batch sizes, so I'm not too worried.

simi commented:
@byroot The different type of query is not related to batch size. And even in smaller batches (1000) the performance could be worse. See following example.

# new

```sql
EXPLAIN DELETE FROM "contract_configurations"
  WHERE "contract_configurations"."id" IN (
    SELECT "contract_configurations"."id" FROM "contract_configurations" WHERE "contract_configurations"."id" > 592825 AND "contract_configurations"."id" <= 593825
);

Delete on contract_configurations  (cost=0.85..3483.98 rows=938 width=12)
  ->  Nested Loop  (cost=0.85..3483.98 rows=938 width=12)
        ->  Index Scan using contract_configurations_pkey on contract_configurations contract_configurations_1  (cost=0.43..1048.07 rows=938 width=10)
              Index Cond: ((id > 592825) AND (id <= 593825))
        ->  Index Scan using contract_configurations_unique_id_and_model_id on contract_configurations  (cost=0.43..2.60 rows=1 width=10)
              Index Cond: (id = contract_configurations_1.id)
```

# old

```sql
EXPLAIN DELETE FROM "contract_configurations"
  WHERE "contract_configurations"."id" IN (592825, 592826, 592827, 592828, 592829, 592830, 592831, 592832, 592833, 592834, 592835, 592836, 592837, 592838, 592839, 592840, 592841, 592842, 592843, 592844, 592845, 592846, 592847, 592848, 592849, 592850, 592851, 592852, 592853, 592854, 592855, 592856, 592857, 592858, 592859, 592860, 592861, 592862, 592863, 592864, 592865, 592866, 592867, 592868, 592869, 592870, 592871, 592872, 592873, 592874, 592875, 592876, 592877, 592878, 592879, 592880, 592881, 592882, 592883, 592884, 592885, 592886, 592887, 592888, 592889, 592890, 592891, 592892, 592893, 592894, 592895, 592896, 592897, 592898, 592899, 592900, 592901, 592902, 592903, 592904, 592905, 592906, 592907, 592908, 592909, 592910, 592911, 592912, 592913, 592914, 592915, 592916, 592917, 592918, 592919, 592920, 592921, 592922, 592923, 592924, 592925, 592926, 592927, 592928, 592929, 592930, 592931, 592932, 592933, 592934, 592935, 592936, 592937, 592938, 592939, 592940, 592941, 592942, 592943, 592944, 592945, 592946, 592947, 592948, 592949, 592950, 592951, 592952, 592953, 592954, 592955, 592956, 592957, 592958, 592959, 592960, 592961, 592962, 592963, 592964, 592965, 592966, 592967, 592968, 592969, 592970, 592971, 592972, 592973, 592974, 592975, 592976, 592977, 592978, 592979, 592980, 592981, 592982, 592983, 592984, 592985, 592986, 592987, 592988, 592989, 592990, 592991, 592992, 592993, 592994, 592995, 592996, 592997, 592998, 592999, 593000, 593001, 593002, 593003, 593004, 593005, 593006, 593007, 593008, 593009, 593010, 593011, 593012, 593013, 593014, 593015, 593016, 593017, 593018, 593019, 593020, 593021, 593022, 593023, 593024, 593025, 593026, 593027, 593028, 593029, 593030, 593031, 593032, 593033, 593034, 593035, 593036, 593037, 593038, 593039, 593040, 593041, 593042, 593043, 593044, 593045, 593046, 593047, 593048, 593049, 593050, 593051, 593052, 593053, 593054, 593055, 593056, 593057, 593058, 593059, 593060, 593061, 593062, 593063, 593064, 593065, 593066, 593067, 593068, 593069, 593070, 593071, 593072, 593073, 593074, 593075, 593076, 593077, 593078, 593079, 593080, 593081, 593082, 593083, 593084, 593085, 593086, 593087, 593088, 593089, 593090, 593091, 593092, 593093, 593094, 593095, 593096, 593097, 593098, 593099, 593100, 593101, 593102, 593103, 593104, 593105, 593106, 593107, 593108, 593109, 593110, 593111, 593112, 593113, 593114, 593115, 593116, 593117, 593118, 593119, 593120, 593121, 593122, 593123, 593124, 593125, 593126, 593127, 593128, 593129, 593130, 593131, 593132, 593133, 593134, 593135, 593136, 593137, 593138, 593139, 593140, 593141, 593142, 593143, 593144, 593145, 593146, 593147, 593148, 593149, 593150, 593151, 593152, 593153, 593154, 593155, 593156, 593157, 593158, 593159, 593160, 593161, 593162, 593163, 593164, 593165, 593166, 593167, 593168, 593169, 593170, 593171, 593172, 593173, 593174, 593175, 593176, 593177, 593178, 593179, 593180, 593181, 593182, 593183, 593184, 593185, 593186, 593187, 593188, 593189, 593190, 593191, 593192, 593193, 593194, 593195, 593196, 593197, 593198, 593199, 593200, 593201, 593202, 593203, 593204, 593205, 593206, 593207, 593208, 593209, 593210, 593211, 593212, 593213, 593214, 593215, 593216, 593217, 593218, 593219, 593220, 593221, 593222, 593223, 593224, 593225, 593226, 593227, 593228, 593229, 593230, 593231, 593232, 593233, 593234, 593235, 593236, 593237, 593238, 593239, 593240, 593241, 593242, 593243, 593244, 593245, 593246, 593247, 593248, 593249, 593250, 593251, 593252, 593253, 593254, 593255, 593256, 593257, 593258, 593259, 593260, 593261, 593262, 593263, 593264, 593265, 593266, 593267, 593268, 593269, 593270, 593271, 593272, 593273, 593274, 593275, 593276, 593277, 593278, 593279, 593280, 593281, 593282, 593283, 593284, 593285, 593286, 593287, 593288, 593289, 593290, 593291, 593292, 593293, 593294, 593295, 593296, 593297, 593298, 593299, 593300, 593301, 593302, 593303, 593304, 593305, 593306, 593307, 593308, 593309, 593310, 593311, 593312, 593313, 593314, 593315, 593316, 593317, 593318, 593319, 593320, 593321, 593322, 593323, 593324, 593325, 593326, 593327, 593328, 593329, 593330, 593331, 593332, 593333, 593334, 593335, 593336, 593337, 593338, 593339, 593340, 593341, 593342, 593343, 593344, 593345, 593346, 593347, 593348, 593349, 593350, 593351, 593352, 593353, 593354, 593355, 593356, 593357, 593358, 593359, 593360, 593361, 593362, 593363, 593364, 593365, 593366, 593367, 593368, 593369, 593370, 593371, 593372, 593373, 593374, 593375, 593376, 593377, 593378, 593379, 593380, 593381, 593382, 593383, 593384, 593385, 593386, 593387, 593388, 593389, 593390, 593391, 593392, 593393, 593394, 593395, 593396, 593397, 593398, 593399, 593400, 593401, 593402, 593403, 593404, 593405, 593406, 593407, 593408, 593409, 593410, 593411, 593412, 593413, 593414, 593415, 593416, 593417, 593418, 593419, 593420, 593421, 593422, 593423, 593424, 593425, 593426, 593427, 593428, 593429, 593430, 593431, 593432, 593433, 593434, 593435, 593436, 593437, 593438, 593439, 593440, 593441, 593442, 593443, 593444, 593445, 593446, 593447, 593448, 593449, 593450, 593451, 593452, 593453, 593454, 593455, 593456, 593457, 593458, 593459, 593460, 593461, 593462, 593463, 593464, 593465, 593466, 593467, 593468, 593469, 593470, 593471, 593472, 593473, 593474, 593475, 593476, 593477, 593478, 593479, 593480, 593481, 593482, 593483, 593484, 593485, 593486, 593487, 593488, 593489, 593490, 593491, 593492, 593493, 593494, 593495, 593496, 593497, 593498, 593499, 593500, 593501, 593502, 593503, 593504, 593505, 593506, 593507, 593508, 593509, 593510, 593511, 593512, 593513, 593514, 593515, 593516, 593517, 593518, 593519, 593520, 593521, 593522, 593523, 593524, 593525, 593526, 593527, 593528, 593529, 593530, 593531, 593532, 593533, 593534, 593535, 593536, 593537, 593538, 593539, 593540, 593541, 593542, 593543, 593544, 593545, 593546, 593547, 593548, 593549, 593550, 593551, 593552, 593553, 593554, 593555, 593556, 593557, 593558, 593559, 593560, 593561, 593562, 593563, 593564, 593565, 593566, 593567, 593568, 593569, 593570, 593571, 593572, 593573, 593574, 593575, 593576, 593577, 593578, 593579, 593580, 593581, 593582, 593583, 593584, 593585, 593586, 593587, 593588, 593589, 593590, 593591, 593592, 593593, 593594, 593595, 593596, 593597, 593598, 593599, 593600, 593601, 593602, 593603, 593604, 593605, 593606, 593607, 593608, 593609, 593610, 593611, 593612, 593613, 593614, 593615, 593616, 593617, 593618, 593619, 593620, 593621, 593622, 593623, 593624, 593625, 593626, 593627, 593628, 593629, 593630, 593631, 593632, 593633, 593634, 593635, 593636, 593637, 593638, 593639, 593640, 593641, 593642, 593643, 593644, 593645, 593646, 593647, 593648, 593649, 593650, 593651, 593652, 593653, 593654, 593655, 593656, 593657, 593658, 593659, 593660, 593661, 593662, 593663, 593664, 593665, 593666, 593667, 593668, 593669, 593670, 593671, 593672, 593673, 593674, 593675, 593676, 593677, 593678, 593679, 593680, 593681, 593682, 593683, 593684, 593685, 593686, 593687, 593688, 593689, 593690, 593691, 593692, 593693, 593694, 593695, 593696, 593697, 593698, 593699, 593700, 593701, 593702, 593703, 593704, 593705, 593706, 593707, 593708, 593709, 593710, 593711, 593712, 593713, 593714, 593715, 593716, 593717, 593718, 593719, 593720, 593721, 593722, 593723, 593724, 593725, 593726, 593727, 593728, 593729, 593730, 593731, 593732, 593733, 593734, 593735, 593736, 593737, 593738, 593739, 593740, 593741, 593742, 593743, 593744, 593745, 593746, 593747, 593748, 593749, 593750, 593751, 593752, 593753, 593754, 593755, 593756, 593757, 593758, 593759, 593760, 593761, 593762, 593763, 593764, 593765, 593766, 593767, 593768, 593769, 593770, 593771, 593772, 593773, 593774, 593775, 593776, 593777, 593778, 593779, 593780, 593781, 593782, 593783, 593784, 593785, 593786, 593787, 593788, 593789, 593790, 593791, 593792, 593793, 593794, 593795, 593796, 593797, 593798, 593799, 593800, 593801, 593802, 593803, 593804, 593805, 593806, 593807, 593808, 593809, 593810, 593811, 593812, 593813, 593814, 593815, 593816, 593817, 593818, 593819, 593820, 593821, 593822, 593823, 593824, 593825
);

Delete on contract_configurations  (cost=0.43..2593.93 rows=1001 width=6)
  ->  Index Scan using contract_configurations_pkey on contract_configurations  (cost=0.43..2593.93 rows=1001 width=6)
        Index Cond: (id = ANY ('{592825,592826,592827,592828,592829,592830,592831,592832,592833,592834,592835,592836,592837,592838,592839,592840,592841,592842,592843,592844,592845,592846,592847,592848,592849,592850,592851,592852,592853,592854,592855,592856,592857,592858,592859,592860,592861,592862,592863,592864,592865,592866,592867,592868,592869,592870,592871,592872,592873,592874,592875,592876,592877,592878,592879,592880,592881,592882,592883,592884,592885,592886,592887,592888,592889,592890,592891,592892,592893,592894,592895,592896,592897,592898,592899,592900,592901,592902,592903,592904,592905,592906,592907,592908,592909,592910,592911,592912,592913,592914,592915,592916,592917,592918,592919,592920,592921,592922,592923,592924,592925,592926,592927,592928,592929,592930,592931,592932,592933,592934,592935,592936,592937,592938,592939,592940,592941,592942,592943,592944,592945,592946,592947,592948,592949,592950,592951,592952,592953,592954,592955,592956,592957,592958,592959,592960,592961,592962,592963,592964,592965,592966,592967,592968,592969,592970,592971,592972,592973,592974,592975,592976,592977,592978,592979,592980,592981,592982,592983,592984,592985,592986,592987,592988,592989,592990,592991,592992,592993,592994,592995,592996,592997,592998,592999,593000,593001,593002,593003,593004,593005,593006,593007,593008,593009,593010,593011,593012,593013,593014,593015,593016,593017,593018,593019,593020,593021,593022,593023,593024,593025,593026,593027,593028,593029,593030,593031,593032,593033,593034,593035,593036,593037,593038,593039,593040,593041,593042,593043,593044,593045,593046,593047,593048,593049,593050,593051,593052,593053,593054,593055,593056,593057,593058,593059,593060,593061,593062,593063,593064,593065,593066,593067,593068,593069,593070,593071,593072,593073,593074,593075,593076,593077,593078,593079,593080,593081,593082,593083,593084,593085,593086,593087,593088,593089,593090,593091,593092,593093,593094,593095,593096,593097,593098,593099,593100,593101,593102,593103,593104,593105,593106,593107,593108,593109,593110,593111,593112,593113,593114,593115,593116,593117,593118,593119,593120,593121,593122,593123,593124,593125,593126,593127,593128,593129,593130,593131,593132,593133,593134,593135,593136,593137,593138,593139,593140,593141,593142,593143,593144,593145,593146,593147,593148,593149,593150,593151,593152,593153,593154,593155,593156,593157,593158,593159,593160,593161,593162,593163,593164,593165,593166,593167,593168,593169,593170,593171,593172,593173,593174,593175,593176,593177,593178,593179,593180,593181,593182,593183,593184,593185,593186,593187,593188,593189,593190,593191,593192,593193,593194,593195,593196,593197,593198,593199,593200,593201,593202,593203,593204,593205,593206,593207,593208,593209,593210,593211,593212,593213,593214,593215,593216,593217,593218,593219,593220,593221,593222,593223,593224,593225,593226,593227,593228,593229,593230,593231,593232,593233,593234,593235,593236,593237,593238,593239,593240,593241,593242,593243,593244,593245,593246,593247,593248,593249,593250,593251,593252,593253,593254,593255,593256,593257,593258,593259,593260,593261,593262,593263,593264,593265,593266,593267,593268,593269,593270,593271,593272,593273,593274,593275,593276,593277,593278,593279,593280,593281,593282,593283,593284,593285,593286,593287,593288,593289,593290,593291,593292,593293,593294,593295,593296,593297,593298,593299,593300,593301,593302,593303,593304,593305,593306,593307,593308,593309,593310,593311,593312,593313,593314,593315,593316,593317,593318,593319,593320,593321,593322,593323,593324,593325,593326,593327,593328,593329,593330,593331,593332,593333,593334,593335,593336,593337,593338,593339,593340,593341,593342,593343,593344,593345,593346,593347,593348,593349,593350,593351,593352,593353,593354,593355,593356,593357,593358,593359,593360,593361,593362,593363,593364,593365,593366,593367,593368,593369,593370,593371,593372,593373,593374,593375,593376,593377,593378,593379,593380,593381,593382,593383,593384,593385,593386,593387,593388,593389,593390,593391,593392,593393,593394,593395,593396,593397,593398,593399,593400,593401,593402,593403,593404,593405,593406,593407,593408,593409,593410,593411,593412,593413,593414,593415,593416,593417,593418,593419,593420,593421,593422,593423,593424,593425,593426,593427,593428,593429,593430,593431,593432,593433,593434,593435,593436,593437,593438,593439,593440,593441,593442,593443,593444,593445,593446,593447,593448,593449,593450,593451,593452,593453,593454,593455,593456,593457,593458,593459,593460,593461,593462,593463,593464,593465,593466,593467,593468,593469,593470,593471,593472,593473,593474,593475,593476,593477,593478,593479,593480,593481,593482,593483,593484,593485,593486,593487,593488,593489,593490,593491,593492,593493,593494,593495,593496,593497,593498,593499,593500,593501,593502,593503,593504,593505,593506,593507,593508,593509,593510,593511,593512,593513,593514,593515,593516,593517,593518,593519,593520,593521,593522,593523,593524,593525,593526,593527,593528,593529,593530,593531,593532,593533,593534,593535,593536,593537,593538,593539,593540,593541,593542,593543,593544,593545,593546,593547,593548,593549,593550,593551,593552,593553,593554,593555,593556,593557,593558,593559,593560,593561,593562,593563,593564,593565,593566,593567,593568,593569,593570,593571,593572,593573,593574,593575,593576,593577,593578,593579,593580,593581,593582,593583,593584,593585,593586,593587,593588,593589,593590,593591,593592,593593,593594,593595,593596,593597,593598,593599,593600,593601,593602,593603,593604,593605,593606,593607,593608,593609,593610,593611,593612,593613,593614,593615,593616,593617,593618,593619,593620,593621,593622,593623,593624,593625,593626,593627,593628,593629,593630,593631,593632,593633,593634,593635,593636,593637,593638,593639,593640,593641,593642,593643,593644,593645,593646,593647,593648,593649,593650,593651,593652,593653,593654,593655,593656,593657,593658,593659,593660,593661,593662,593663,593664,593665,593666,593667,593668,593669,593670,593671,593672,593673,593674,593675,593676,593677,593678,593679,593680,593681,593682,593683,593684,593685,593686,593687,593688,593689,593690,593691,593692,593693,593694,593695,593696,593697,593698,593699,593700,593701,593702,593703,593704,593705,593706,593707,593708,593709,593710,593711,593712,593713,593714,593715,593716,593717,593718,593719,593720,593721,593722,593723,593724,593725,593726,593727,593728,593729,593730,593731,593732,593733,593734,593735,593736,593737,593738,593739,593740,593741,593742,593743,593744,593745,593746,593747,593748,593749,593750,593751,593752,593753,593754,593755,593756,593757,593758,593759,593760,593761,593762,593763,593764,593765,593766,593767,593768,593769,593770,593771,593772,593773,593774,593775,593776,593777,593778,593779,593780,593781,593782,593783,593784,593785,593786,593787,593788,593789,593790,593791,593792,593793,593794,593795,593796,593797,593798,593799,593800,593801,593802,593803,593804,593805,593806,593807,593808,593809,593810,593811,593812,593813,593814,593815,593816,593817,593818,593819,593820,593821,593822,593823,593824,593825}'::integer[]))
```

# best of both worlds - probably most efficient (would need better logic of select -> delete transform in AR)

```sql
EXPLAIN DELETE FROM "contract_configurations" WHERE "contract_configurations"."id" > 592825 AND "contract_configurations"."id" <= 593825;

Delete on contract_configurations  (cost=0.43..1048.07 rows=938 width=6)
  ->  Index Scan using contract_configurations_pkey on contract_configurations  (cost=0.43..1048.07 rows=938 width=6)
        Index Cond: ((id > 592825) AND (id <= 593825))
```

simi commented:
> I read the diff twice and I still don't where this nested querying is coming from. Is this query outdated or I'm just tired?

I have explained that in https://github.com/rails/rails/pull/45414#issuecomment-1168980701.

> Anyway I'm not really familiar with in_batches feature and real usage. I doubt Model.in_batches(of: 100_000) {|b| b.load} is close to any useful code. What's the original usecase for in_batches? Looking at ActiveRecord::Batches::BatchEnumerator it seems it is useful for `delete_all`, `destroy_all` and `update_all`.

Those examples are based on `delete_all` usage (https://api.rubyonrails.org/classes/ActiveRecord/Batches/BatchEnumerator.html#method-i-delete_all).

simi commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -201,15 +208,15 @@ def find_in_batches(start: nil, finish: nil, batch_size: 1000, error_on_ignore:
     # NOTE: By its nature, batch processing is subject to race conditions if
     # other processes are modifying the database.
-    def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore: nil, order: :asc)
+    def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore: nil, order: :asc, use_ranges: nil)
```
What about to make it more generic (extendable) in the future? I'm still trying to find out way to introduce PG cursors to ActiveRecord. This could be the way. What about `scope_by: :ids` by default and `scope_by: :range` with this optional addition? I can later try to contribute `scope_by: :cursor`.

:thinking: _I'm really bad at naming, I believe there is better name for this argument._

fatkodima replied:
Do I correctly understand, that cursors can be used only when `load: true` is passed?

simi replied:
No, why? You can move in iteration using `MOVE` when you don't ask for the data to be loaded.  And even it would be needed, what's the problem?

fatkodima commented:
@simi Can you ran the actual queries to prove the new changes can be slower? Or can you provide a test script to run by myself? Query plans are nice, but the actual execution can show different timings.

fatkodima commented:
@simi I updated this PR. Now it does not generate a subquery for updates/deletes and is actually faster for any batch size (so the previous problem disappeared).

I also updated all my previous comments with measurements with the updated numbers, which are way better now than before.

simi commented:
I was looking why the subquery is selected for given queries (delete_all, update_all) and the reason is actually `LIMIT` clause being applied. Would it make sense to remove the `LIMIT` clause actually? When passing array of IDs (primary keys), the limit is redundant (since the max amount of rows is amount of IDs). When passing range (this PR) it is subject of race conditions anyway.

simi commented:
> @simi I updated this PR. Now it does not generate a subquery for updates/deletes and is actually faster for any batch size (so the previous problem disappeared).
> 
> I also updated all my previous comments with measurements with the updated numbers, which are way better now than before.

Ahh, I think we got to the same point. :muscle: Let me check.

byroot commented:
I tested it locally, the sub queries are gone, so I'm good with this. Merging.

simi commented on activerecord/lib/active_record/relation/batches.rb:
```diff
@@ -226,13 +231,22 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
+          finish = ids.last
+          if finish
+            yielded_relation = apply_finish_limit(batch_relation, finish, order)
+            yielded_relation = yielded_relation.except(:limit, :order)
```
If I understand it well, this line is now crucial in here and it is missing currently test coverage (I tried to comment it out and tests were still passing).

Would it be possible to add at least comment and test case? Originally I was thinking to move this logic (excluding `LIMIT` and `ORDER` clauses) to actual `ActiveRecord::Batches::BatchEnumerator` since it is useful for methods in there.

fatkodima replied:
I believe the `ActiveRecord::Batches::BatchEnumerator` should not modify the relation. 
But for comment and tests - 👍 Please, open a follow up PR if you would like to.

simi replied:
resolved by https://github.com/rails/rails/pull/45671 :muscle:

jrochkind commented:
What's going to happen if the table has UUID pks?

byroot commented:
> What's going to happen if the table has UUID pks?

As noted in the documentation, the `in_batches` family of methods requires an orderable primary key:

> NOTE: Order can be ascending (:asc) or descending (:desc). It is automatically set to ascending on the primary key (“id ASC”). This also means that this method only works when the primary key is orderable (e.g. an integer or string).

jrochkind commented:
> As noted in the documentation, the in_batches family of methods requires an orderable primary key:

Interesting. I don't recall noticing this before, but I have in fact  used `find_each` and `in_batches` on tables with a UUID key in postgres.  It appeared to work! 

It looks like this works because UUIDs are in fact orderable in postgres!  I can do an SQL with `ORDER BY id ASC` or `DESC` with a UUID `id` no problem, and the order is deterministic. 

So I guess if this implementation can already work with strings as well as integers, it will also work with UUID in a database where they are orderable?
